cub/block/block_run_length_decode.cuh

```diff
@@ -0,0 +1,435 @@
+      thread_dst_offset++;
+    }
+
+    // Write the runs' total decoded size into the last element
+    if (linear_tid == 0)
+    {
+      temp_storage.runs.run_offsets[BLOCK_RUNS] = total_decoded_size;
+    }
+
+    // Ensure run offsets have been writen to shared memory
+    __syncthreads();
+  }
+
+  /**
+   * @brief Run-length decodes the runs previously passed via a call to Init(...) and returns the run-length decoded
+   * items in a blocked arrangement to \p decoded_items. If the number of run-length decoded items exceeds the
+   * run-length decode buffer (i.e., <b>DECODED_ITEMS_PER_THREAD * BLOCK_THREADS</b>), only the items that fit within
+   * the buffer are returned. Subsequent calls to <b>RunLengthDecode</b> adjusting \p from_decoded_offset can be
+   * used to retrieve the remaining run-length decoded items.
+   *
+   * @param from_decoded_offset If invoked with from_decoded_offset that is larger than total_decoded_size results in
+   * undefined behavior.
+   */
+  template <typename UserRelativeOffsetT>
+  __device__ __forceinline__ void RunLengthDecode(UniqueItemT (&decoded_items)[DECODED_ITEMS_PER_THREAD],
+                                                  UserRelativeOffsetT (&item_offsets)[DECODED_ITEMS_PER_THREAD],
+                                                  int32_t from_decoded_offset = 0)
+  {
+    // Populate run-length decode buffer with continuation-of-current-run items
+    InitDecodeBuffer(temp_storage.runs.decode_buffer);
+
+    // Ensure that the continuation-of-current-run items have been written to shared memory
+    __syncthreads();
+
+    // Scatter the beginning of each run into the run-length decode buffer
+    RunOffsetT run_offset = linear_tid;
+#pragma unroll
+    for (uint32_t i = 0; i < RUNS_PER_THREAD; i++)
+    {
+      // Fetch this thread's run and subsequently check whether it falls into this pass' run-length decode buffer
+      int32_t run_target_idx      = temp_storage.runs.run_offsets[run_offset];
+      int32_t next_run_target_idx = temp_storage.runs.run_offsets[run_offset + 1];
+      int32_t run_length          = next_run_target_idx - run_target_idx;
+
+      // Whether this run begins before the from_decoded_offset but has trailing items that still fall within the
+      // beginning of this buffer
+      bool is_a_trailing_run = (run_target_idx < from_decoded_offset) && (next_run_target_idx > from_decoded_offset);
+
+      // In particular for the trailing run, we want to scatter into index 0 (not a negative scatter index)
+      int32_t scatter_target_idx = max(0, run_target_idx - from_decoded_offset);
+      int32_t relative_offset    = is_a_trailing_run ? from_decoded_offset - run_target_idx : 0;
+
+      // Only write this beginning-of-run if the run length falls within the decoded buffer OR
+      // the run is a trailing run.
+      if (run_length > 0 && (run_target_idx >= from_decoded_offset || is_a_trailing_run) &&
```

Comment:
We try to avoid using `&&` in a condition statement for flow control in CUB. Could you use nested if statements instead? The style guide is being developed 😄

Reply:
This has been removed with the switch to the binary search-based method. But I tried to adhere to it in the tests too 👍